pkg/controller/job/job_controller.go

```diff
@@ -456,34 +492,28 @@ func (jm *JobController) syncJob(key string) error {
 			jm.queue.AddAfter(key, time.Duration(*job.Spec.ActiveDeadlineSeconds)*time.Second)
 		}
 	}
-	// if job was finished previously, we don't want to redo the termination
-	if IsJobFinished(&job) {
-		return nil
-	}
 
 	var manageJobErr error
-	if pastActiveDeadline(&job) {
-		// TODO: below code should be replaced with pod termination resulting in
-		// pod failures, rather than killing pods. Unfortunately none such solution
-		// exists ATM. There's an open discussion in the topic in
-		// https://github.com/kubernetes/kubernetes/issues/14602 which might give
-		// some sort of solution to above problem.
-		// kill remaining active pods
-		wait := sync.WaitGroup{}
-		errCh := make(chan error, int(active))
-		wait.Add(int(active))
-		for i := int32(0); i < active; i++ {
-			go func(ix int32) {
-				defer wait.Done()
-				if err := jm.podControl.DeletePod(job.Namespace, activePods[ix].Name, &job); err != nil {
-					defer utilruntime.HandleError(err)
-					glog.V(2).Infof("Failed to delete %v, job %q/%q deadline exceeded", activePods[ix].Name, job.Namespace, job.Name)
-					errCh <- err
-				}
-			}(i)
-		}
-		wait.Wait()
-
+	jobFailed := false
+	var failureReason string
+	var failureMessage string
+
+	jobHaveNewFailure := failed > job.Status.Failed
+
+	// check if the number of failed jobs increased since the last syncJob
+	if jobHaveNewFailure && (previousRetry+1 > *job.Spec.BackoffLimit) {
```

Comment:
Using it for this?  I'd keep track of my retries locally to start.  Of a few dozen uses, you've found one that want this feature and its not obvious to me that all backoffs are created equal.  

Also, does `AddRateLimited` on the workqueue not do what you want?

Reply:
Looking at others controllers: ```federated.JobController``` and the ```Kubelet``` the usage of the ```flowcontrol.Backoff ``` was looking the best choice,

Adding a separated data-structure for storing the number of retries for each Job was a possibility that I envisaged. 
But the fact that we still need to update the ```Backoff``` structure in order to allow to set different initial delay durations for each entry, made me orient myself to integrate also the ```retry```counter in the ```Backoff``` struct